libs/community/langchain_community/document_loaders/batch_loader.py

```diff
@@ -0,0 +1,288 @@
+    return a tqdm iterator.
+
+    Args:
+        iterable (Iterable[T]): any iterable
+        args_length (int): length of the iterable
+        show_progress (bool, optional): If true, will return a tqdm iterator.
+            Defaults to False.
+
+    Raises:
+        ImportError: if tqdm is not installed and show_progress is True
+    """
+    if show_progress:
+        try:
+            from tqdm import tqdm
+        except ImportError as err:
+            raise ImportError(
+                "You must install tqdm to use show_progress=True."
+                "You can install tqdm with `pip install tqdm`."
+            ) from err
+        return list(tqdm(iterable, total=args_length))
+    return list(iterable)
+
+
+def _load_loader(
+    loader_args: Dict[str, Any], loader_callable: Callable[..., BaseLoader]
+) -> List[Document]:
+    """Function used to instantiate the loader and load the data.
+
+    Args:
+        loader_args (Dict[str, Any]): list of loader_args used
+            to initialize the loader. The keys of the dict are the name of
+            the arguments and the values are the list of values for each.
+        loader_callable (Callable[..., BaseLoader]): callable
+            used to initialize the loader (class or init function )
+
+    Returns:
+        List[Document]: list of loaded documents
+    """
+    loader = loader_callable(**loader_args)
+    return loader.load()
+
+
+Methods = Literal["sequential", "thread", "process", "async"]
+
+
+class BatchLoader(BaseLoader):
+    """
+    Meta loader use to load multiple data from a loader either in parallel
+    or sequentially.
+
+    Example:
+
+    Using a class that is instantiate with __init__
+    ```
+    file_path_list : List[str] = [...]
+    loader_files = BatchLoader(
+        TextLoader,
+        {
+            "file_path": file_path_list
+        },
+        method="process",
+        max_workers=4,
+    )
+    ```
+    Using a static method to initialize the loader:
+    ```
+    youtube_video_urls : List[str] = [...]
+    loader_youtube = BatchLoader(
+        YoutubeLoader.from_url,
+        {
+            "url": youtube_video_urls
+        },
+    )
+    ```
+
+    Then you can load your data:
+    ```
+    documents = loader_files.load()
+    ```
+    """
+
+    def __init__(
+        self,
+        loader_initializer_callable: Callable[..., BaseLoader],
+        init_loader_args: Dict[str, List[Any]],
+        *,
+        method: Methods = "sequential",
+        max_workers: int = 1,
+        show_progress: bool = False,
+    ):
+        """Initialize the loader with a loader_callable (either a class or a
+        function used to initialize the loader) and a list of loader_args.
+
+        Possible methods are:
+            - "sequential": load data sequentially
+            - "thread": load data using threads
+            - "process": load data using processes
+            - "async": load data using asyncio
```

Comment:
`async` doesn't seem needed as this mode is determined by the usage of `load` vs `aload`.
`sequential` could also be removed by assuming that if the concurrency is 1, then it will be sequential (can detect that if max_workers is 1 then there's no need to spawn a thread-pool or processes and the operation can be done in the current thread).

Reply:
You're correct. It might be better to have only the max_concurrency argument and then call either alazy_load or lazy_load.

thread isn't useful at all. Perhaps in the future with subinterpreter, there could be some interest, but for now, let's get rid of it.